pkg/daemon/daemon.go

```diff
@@ -888,29 +901,50 @@ func (dn *Daemon) handleNodeUpdate(old, cur interface{}) {
 // prepUpdateFromCluster handles the shared update prepping functionality for
 // flows that expect the cluster to already be available. Returns true if an
 // update is required, false otherwise.
-func (dn *Daemon) prepUpdateFromCluster() (bool, error) {
+func (dn *Daemon) prepUpdateFromCluster() (*mcfgv1.MachineConfig, *mcfgv1.MachineConfig, error) {
 	desiredConfigName, err := getNodeAnnotationExt(dn.node, constants.DesiredMachineConfigAnnotationKey, true)
 	if err != nil {
-		return false, err
+		return nil, nil, err
+	}
+	desiredConfig, err := dn.mcLister.Get(desiredConfigName)
+	if err != nil {
+		return nil, nil, err
 	}
 	// currentConfig is always expected to be there as loadNodeAnnotations
 	// is one of the very first calls when the daemon starts.
 	currentConfigName, err := getNodeAnnotation(dn.node, constants.CurrentMachineConfigAnnotationKey)
 	if err != nil {
-		return false, err
+		return nil, nil, err
+	}
+	currentConfig, err := dn.mcLister.Get(currentConfigName)
+	if err != nil {
+		return nil, nil, err
 	}
 	state, err := getNodeAnnotation(dn.node, constants.MachineConfigDaemonStateAnnotationKey)
 	if err != nil {
-		return false, err
+		return nil, nil, err
+	}
+
+	mcJSON, err := os.Open(currentConfigFingerprint)
```

Comment:
Don't we need to do this *before* we load `currentConfig`?

Reply:
uhm, we may, I can't see why though, would you explain (I feel dumb)

Reply:
If the current config isn't available in the apiserver, won't the `Get` above fail before we can find our cached version?

Reply:
oh gosh yeah

Reply:
No wait, the scenario here is that after a restore we still have something in the apiserver, which is really what we want on the machine itself, if current isn't on the apiserver, there's not much we can do anyway (which is the BZ we got from David yesterday but this PR isn't tackling that).

This is something I wrote to understand this better:
```
- current=configA == desired=configA
upgrade
- current=configA != desired=configB (desired being generated by upgrade and it's being applied as new current)
roll out new configs and MCD syncs
- current=configB == desired=configB

...

trigger a backup restore pre-upgrade which isn't calling informers
- current=configA == desired=configA
but now what's on disk is still configB

we now need something which retriggers the MCD to sync to configA
```

`configA` is what we need anyway, so it must be in the apiserver after a restore

Reply:
can i ask: when is a backup restore triggered? i feel like i understand what this pr is supposed to do, but not fully why?

Reply:
there's work still ongoing on others to spec out how/when/where an etcd backup is performed (and a subsequent restore)

Reply:
@cgwalters can you check my answer above? I think this is working as intended but I'm waiting on a way to test this out